File: app/src/main/java/dev/sergei/miniwebserver/domain/repository/StorageRepository.kt
package dev.sergei.miniwebserver.domain.repository

import dev.sergei.miniwebserver.domain.model.DirListing
import dev.sergei.miniwebserver.domain.model.Folder
import dev.sergei.miniwebserver.domain.model.OpenFile
import dev.sergei.miniwebserver.domain.model.SearchResult
import java.io.InputStream

interface StorageRepository {
    fun getFolders(): List<Folder>

    fun addFolder(folderId: String)

    fun removeFolder(folderId: String)

    fun list(
        folderId: String,
        path: List<String>,
    ): DirListing

    fun search(
        folderId: String,
        query: String,
    ): SearchResult

    fun createDirectory(
        folderId: String,
        path: List<String>,
        name: String,
    )

    fun delete(
        folderId: String,
        path: List<String>,
        name: String,
    )

    fun exists(
        folderId: String,
        path: List<String>,
        name: String,
    ): Boolean

    fun open(
        folderId: String,
        path: List<String>,
        name: String,
    ): OpenFile

    fun upload(
        folderId: String,
        path: List<String>,
        name: String,
        source: InputStream,
        overwrite: Boolean,
    )
}

File: web/src/domain/repositories/StorageRepository.ts
import type { ActionResult, DirListing, Folder, SearchHit } from '@/domain/models/types'

export interface StorageRepository {
  getFolders(): Promise<Folder[]>
  list(folderId: string, path: string[]): Promise<DirListing>
  search(folderId: string, query: string): Promise<SearchHit[]>
  createDirectory(folderId: string, path: string[], name: string): Promise<ActionResult>
  deleteEntry(folderId: string, path: string[], name: string): Promise<ActionResult>
  exists(folderId: string, path: string[], name: string): Promise<boolean>
  downloadUrl(folderId: string, path: string[], name: string): string
  uploadFile(
    folderId: string,
    path: string[],
    file: File,
    overwrite: boolean,
    onProgress?: (fraction: number) => void,
  ): Promise<ActionResult>
}
